File: drills/ts/solutions/nc69_palindromic_substrings.ts
export function countSubstrings(s: string): number {
  // Same 2n centres as finding the longest one, except that here every
  // successful widening is itself an answer, so the count is how many times the
  // expansion succeeded rather than how far it got.
  let total = 0;
  for (let i = 0; i < s.length; i++) total += grow(s, i, i) + grow(s, i, i + 1);
  return total;
}

function grow(s: string, left: number, right: number): number {
  let count = 0;
  while (left >= 0 && right < s.length && s[left] === s[right]) {
    count++;
    left--;
    right++;
  }
  return count;
}

File: drills/go/solutions/nc69_palindromic_substrings.go
package main

func countSubstrings(s string) int {
	count := 0
	// Expand from each of the 2n-1 centres; every successful widening is
	// one more palindrome.
	for centre := 0; centre < len(s); centre++ {
		count += expandCount(s, centre, centre) + expandCount(s, centre, centre+1)
	}
	return count
}

func expandCount(s string, left, right int) int {
	count := 0
	for left >= 0 && right < len(s) && s[left] == s[right] {
		count++
		left--
		right++
	}
	return count
}
